webapp/lib/controller_tag_helper.dart

```diff
@@ -0,0 +1,59 @@
+part of controller;
+
+class TagManager {
+  static final TagManager _singleton = TagManager._internal();
+
+  List<model.Tag> _tags = <model.Tag>[];
+  List<model.Tag> get tags => _tags;
+
+  TagManager._internal();
+
+  factory TagManager() => _singleton;
+
+  Set<String> get groups => new Set.from(emptyGroups)..addAll(_groupTagsIntoCategories(tags).keys);
+
+  Set<String> emptyGroups = {};
```

Comment:
is this namesOfEmptyGroups? If this is always an empty set, it would be better as a const propagated into the expression

Reply:
first option is what I had in mind - renamed now, thanks!